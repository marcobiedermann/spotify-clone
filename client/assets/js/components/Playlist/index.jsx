import PropTypes from 'prop-types';
import React from 'react';
import { Link } from 'react-router-dom';

const Playlist = ({
  id,
  images,
  name,
  owner,
}) => (
  <figure>
    {images[0] && (
      <Link to={`/users/${owner.id}/playlists/${id}`}>
        <img src={images[0].url} alt={name} />
      </Link>
    )}
    <figcaption>
      <Link to={`/users/${owner.id}/playlists/${id}`}>
        {name}
      </Link>
    </figcaption>
  </figure>
);

Playlist.propTypes = {
  id: PropTypes.string,
  images: PropTypes.arrayOf(PropTypes.shape()),
  name: PropTypes.string,
  owner: PropTypes.shape({
    id: PropTypes.string,
  }),
};

Playlist.defaultProps = {
  id: '',
  images: [],
  name: '',
  owner: {
    id: '',
  },
};

export default Playlist;
