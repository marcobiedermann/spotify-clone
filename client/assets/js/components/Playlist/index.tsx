/* eslint-disable camelcase */

import React, { FC } from 'react';
import { Link } from 'react-router-dom';
import Image, { ImageProps } from '../Image';
import { TracksProps } from '../Tracks';

interface OwnerProps {
  display_name: string;
  id: string;
}

export interface PlaylistProps {
  id: string;
  images: ImageProps[];
  name: string;
  owner: OwnerProps;
  tracks: TracksProps;
}

const Playlist: FC<PlaylistProps> = (props) => {
  const { id, images, name } = props;

  return (
    <figure>
      {images[0] && (
        <Link to={`/playlists/${id}`}>
          <Image {...images[0]} alt={name} />
        </Link>
      )}
      <figcaption>
        <Link to={`/playlists/${id}`}>{name}</Link>
      </figcaption>
    </figure>
  );
};

export default Playlist;
