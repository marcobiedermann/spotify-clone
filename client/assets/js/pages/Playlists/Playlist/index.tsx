import React, { FC } from 'react';
import { Helmet } from 'react-helmet';
import { Link, RouteChildrenProps, useParams } from 'react-router-dom';
import useSWR from 'swr';
import Button from '../../../components/Button';
import Error from '../../../components/Error';
import Loader from '../../../components/Loader';
import Track from '../../../components/Track';

interface Params {
  playlistId: string;
}

const PlaylistPage: FC<RouteChildrenProps> = () => {
  const { playlistId } = useParams<Params>();
  const { data, error } = useSWR(`/v1/playlists/${playlistId}`);

  if (error) {
    return <Error>{error.message}</Error>;
  }

  if (!data) {
    return <Loader />;
  }

  const { name, owner, tracks } = data;

  return (
    <>
      <Helmet>
        <title>Playlist</title>
      </Helmet>
      <div>
        <p>Playlist</p>
        <h1>{name}</h1>
        <p>
          Created by <Link to={`/users/${owner.id}`}>{owner.display_name}</Link> Â· {tracks.total}{' '}
          songs
        </p>
        <p>
          <Button>Play</Button>
        </p>
        <table>
          <tbody>
            {tracks.items.map((track) => (
              <Track key={track.track.id} {...track.track} />
            ))}
          </tbody>
        </table>
      </div>
    </>
  );
};

export default PlaylistPage;
