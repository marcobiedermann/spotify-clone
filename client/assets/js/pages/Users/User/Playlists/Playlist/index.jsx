import React from 'react';

const PlaylistPage = () => (
  <div>
    Playlist
  </div>
);

export default PlaylistPage;
