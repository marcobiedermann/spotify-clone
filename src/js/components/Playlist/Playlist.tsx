import React from 'react';
import { Link } from 'react-router-dom';
import Image from '../Image';

interface Image {
  height: number;
  url: string;
  width: number;
}

export interface PlaylistProps {
  id: string;
  images: Image[];
  name: string;
}

function Playlist(props: PlaylistProps): JSX.Element {
  const { id, images, name } = props;

  return (
    <figure>
      {images[0] && (
        <Link to={`/playlists/${id}`}>
          <Image {...images[0]} alt={name} />
        </Link>
      )}
      <figcaption>
        <Link to={`/playlists/${id}`}>{name}</Link>
      </figcaption>
    </figure>
  );
}

export default Playlist;
