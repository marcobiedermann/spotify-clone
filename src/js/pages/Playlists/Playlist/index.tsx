import { formatDistance } from 'date-fns';
import React from 'react';
import { Helmet } from 'react-helmet';
import { Link, useParams } from 'react-router-dom';
import {
  Button,
  Error,
  Image,
  Loader,
  Media,
  MediaBody,
  MediaObject,
  PlaylistTracks,
  Section,
} from '../../../components';
import { usePlaylist } from '../../../hooks/playlists';

const defaultImageSize = 240;

function PlaylistPage(): JSX.Element {
  const { playlistId } = useParams();
  const { data, error, isError, isPending } = usePlaylist(playlistId!);

  if (isError) {
    return <Error>{error.message}</Error>;
  }

  if (isPending) {
    return <Loader />;
  }

  const { description, images = [], name, owner, tracks } = data;
  const image = images[0];

  const totalDuration =
    tracks?.items.reduce(
      (accumulator, currentValue) => accumulator + (currentValue.track?.duration_ms || 0),
      0,
    ) || 0;

  return (
    <>
      <Helmet>
        <title>Playlist</title>
      </Helmet>
      <div>
        <header>
          <Media>
            <MediaObject>
              <Image
                {...image}
                alt={name || ''}
                width={defaultImageSize}
                height={defaultImageSize}
              />
            </MediaObject>
            <MediaBody>
              <h3>Playlist</h3>
              <h1>{name}</h1>
              {description && <p>{description}</p>}
              <p>
                <strong>
                  <Link to={`/users/${owner?.id}`}>{owner?.display_name}</Link> Â· {tracks?.total}{' '}
                  songs,
                </strong>{' '}
                {formatDistance(0, totalDuration)}
              </p>
            </MediaBody>
          </Media>
        </header>
        <Section>
          <p>
            <Button>Play</Button>
          </p>
        </Section>
        {tracks && (
          <Section>
            <PlaylistTracks items={tracks.items} />
          </Section>
        )}
      </div>
    </>
  );
}

export default PlaylistPage;
