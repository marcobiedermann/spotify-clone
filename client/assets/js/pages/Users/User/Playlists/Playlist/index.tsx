import React, { Component } from 'react';
import { Helmet } from 'react-helmet';
import { connect } from 'react-redux';
import { Link, RouteChildrenProps, useParams } from 'react-router-dom';
import { bindActionCreators } from 'redux';
import { fetchUserPlaylist } from '../../../../../actions/users';
import Button from '../../../../../components/Button';
import Error from '../../../../../components/Error';
import Loader from '../../../../../components/Loader';
import Playlist, { PlaylistProps } from '../../../../../components/Playlist';
import Track from '../../../../../components/Track';

export interface PlaylistPage extends RouteChildrenProps {
  accessToken: string;
  error: {
    message: string;
  };
  fetchUserPlaylist: any;
  isLoading: boolean;
  playlist: PlaylistProps;
}

export class PlaylistPage extends Component<PlaylistPage> {
  componentDidMount() {
    const { accessToken, fetchUserPlaylist } = this.props;
    const { playlistId } = useParams();

    fetchUserPlaylist(accessToken, '11168662039', playlistId);
  }

  render() {
    const { error, isLoading, playlist } = this.props;

    if (error) {
      return <Error>{error.message}</Error>;
    }

    if (isLoading) {
      return <Loader />;
    }

    return (
      <>
        <Helmet>
          <title>Playlist</title>
        </Helmet>
        <div>
          <Playlist {...playlist} />
          <p>Playlist</p>
          <h1>{playlist.name}</h1>
          {playlist.owner && (
            <p>
              Created by{' '}
              <Link to={`/users/${playlist.owner.id}`}>{playlist.owner.display_name}</Link> Â·{' '}
              {playlist.tracks.total} songs
            </p>
          )}
          <p>
            <Button>Play</Button>
          </p>
          {playlist.tracks && (
            <table>
              <tbody>
                {playlist.tracks.items.map((track) => (
                  <Track key={track.id} {...track} />
                ))}
              </tbody>
            </table>
          )}
        </div>
      </>
    );
  }
}

const mapStateToProps = (state) => ({
  ...state,
  error: state.users.error,
  isLoading: state.users.isLoading,
  playlist: state.users.playlist,
});

const mapDispatchToProps = (dispatch) =>
  bindActionCreators(
    {
      fetchUserPlaylist,
    },
    dispatch,
  );

export default connect(mapStateToProps, mapDispatchToProps)(PlaylistPage);
