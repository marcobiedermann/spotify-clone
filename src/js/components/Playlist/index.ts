import Playlist from './Playlist';

export default Playlist;
export * from './Playlist';
