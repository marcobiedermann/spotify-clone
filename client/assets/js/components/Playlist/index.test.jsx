import React from 'react';
import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';
import Playlist from '.';

it('renders correctly', () => {
  const tree = renderer
    .create(
      <MemoryRouter>
        <Playlist />
      </MemoryRouter>,
    )
    .toJSON();

  expect(tree).toMatchSnapshot();
});
