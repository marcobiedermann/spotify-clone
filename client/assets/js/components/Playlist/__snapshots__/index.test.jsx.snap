// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<figure>
  <figcaption>
    <a
      href="/users//playlists/"
      onClick={[Function]}
    >
      
    </a>
  </figcaption>
</figure>
`;
